Samples: Bluetooth: Vendor Specific: support for HCI Scan Request

giansta opened the pull request:
Example of usage of HCI vendor specific Set Scan Request Reports command, Scan Request Received Event and corresponding callback to get application events.

The sample application shows usage as an alternative to extended advertisement saving precious RAM.

With sample application the RAM saving as an alternative to extended advertisement is about 1.5 kB RAM. Building more complex applications gives larger RAM usage saving, around 5 kB.

This PR depends on previous #66529

cvinayak commented:
Tested the changes here. Controller changes already reviewed here: 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/6350656/e3fdc122-d2fa-4324-bfff-f699879379c6)

Tested the sample in this PR:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/6350656/6308a917-2628-4583-a38e-924bd50d4a91)

cvinayak commented on samples/bluetooth/hci_vs_scan_req/prj.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_LOG=y
+CONFIG_BT_BROADCASTER=y
+CONFIG_BT_DEVICE_NAME="VS Scan Notify"
+CONFIG_BT_CTLR_VS_SCAN_REQ_RX=y
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL=y
```
These should be placed in a separate overlay file, say `overlay-bt_ll_sw_split.conf` and have a `sample.yaml` file that be by CI to test building this sample (take a look at other samples that do something similar).

giansta replied:
I pushed you change proposal. The sample app builds on my machine. I don't know why it's getting this error in CI:
`FileNotFoundError: [Errno 2] No such file or directory: '_test_plan_partial.json'`
When I try running on my console:
`scripts/twister -c -T tests/bluetooth/init -T samples/bluetooth/hci_vs_scan_req --all --save-tests _test_plan_partial.json`
I see no errors, yet not sure if it's done correctly.

cvinayak commented:
> Handle vendor specific Scan Request Received Event and provide a callback to get application events.
> 
> A sample application shows usage as an alternative to extended advertisements saving precious 5.5 kB RAM.
> 
> This PR depends on previous #66529

@giansta With the sample in this PR, with `CONFIG_BT_PERIPHERAL=y` and Dynamic Pwr Control disabled:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/6350656/6a3df8cf-9fa2-4194-8eaf-743c322fb21f)


I do not see your upto ~5.5 KB saving in RAM... is your saving in your product application?

jori-nordic commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,312 @@
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->enable = (uint8_t) enable;
+
+	err = bt_hci_cmd_send(BT_HCI_OP_VS_SET_SCAN_REQ_REPORTS, buf);
```
It seems responsibility for this feature is partially implemented in the user's app and in the host: User sends an HCI command manually, and gets a callback from the host.

It needs to be either fully inside or fully outside of the host.

Speaking as a host maintainer, the less code I have to maintain, the happier I am.
You can have it outside by using [`CONFIG_BT_HCI_VS_EVT_USER`](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/Kconfig#L1103).

giansta replied:
Hi @jori-nordic, thanks for your feedback.
I tried using `CONFIG_BT_HCI_VS_EVT_USER` and adding a callback for HCI Vendor-Specific Events in place of the host callback I added.
I see a problem here as the APIs to get the advertiser ID seem to be host internal and not visible at application level.
I'm talking about usage of `adv_get_legacy()` that is made in `bt_hci_le_vs_scan_req_received()`, which is coded similarly to `bt_hci_le_scan_req_received()`.
(BTW I see now an improvement in its last version that I can also add.)
Do you see a smarter way or is this really a blocking point?

jori-nordic replied:
I think there is one:
The only thing we use `adv` for is for getting the identity that started advertising. Since you (the application) started advertising, then you know what identity address you have. 
Correct me if I'm wrong also, but this is not extended advertising right, so we don't have multiple adv sets running.

giansta replied:
Indeed the application can be built both using extended (one set only) and legacy advertising, to compare performances and memory usage. At moment I'm using only legacy adv and I'm patching with local declaration of `bt_le_adv_lookup_legacy`, as it seems that I cannot include `adv.h`, but perhaps we can even not use it.
I've also (temporarily) made public not only `bt_lookup_id_addr`, but even `bt_addr_le_is_resolved` and `bt_addr_le_copy_resolved` that are all moved in `addr.h`.
As a test, I've added the new vs_scanned that is used via `CONFIG_BT_HCI_VS_EVT_USER` and returns always false by purpose to have even the previous callback to compare if they're doing the same. Yet the results are not
totally expected and I'd like to have your opinion about what I'm doing wrong.
1. Different output for the to callbacks. The logs I get are like these ones:
```
00> vs_scanned peer 2D:C6:3C:B4:A2:01 (0x04) rssi 97
00> vs_scanned (address 2D:C6:3C:B4:A2:01 (0x04) rssi 97)
00> [00:00:17.117,340] <err> bt_hci_core: subevent 0x04
00> [00:00:17.117,431] <err> bt_adv: bt_hci_le_vs_scan_req_received peer 61:2D:C6:3C:B4:A2 (random) rssi -127 adv 0x20000018
00> scanned (address 61:2D:C6:3C:B4:A2 (random) rssi -127)
```
The first two line are from the new callback. The data taken from the passed buffers seem incorrect and  the address shifted by one byte between the two functions and the RSSI seems taken from two different parts of the buffer but both wrong.

2. In the new callback vs_scanned the operations managing the address seem redundant, as both logs before and after them log the same values (unless fixing the above point makes a difference). In such case we could remove usage of the APIs made temporarily public.

jori-nordic replied:
could you push the code to another branch and post a link?

giansta replied:
I've pushed it into this branch

jori-nordic replied:
You were missing a `pull_u8` compared to the vs handler: https://github.com/jori-nordic/zephyr/blob/68659-bsim/samples/bluetooth/hci_vs_scan_req/src/main.c#L198-L201

hci_core [pulls that byte](https://github.com/jori-nordic/zephyr/blob/68659-bsim/subsys/bluetooth/host/hci_core.c#L2466-L2469) right before handing control over to the in-stack vs-handler.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/68428270/ac5614b9-e70a-4f1d-876b-780a29e49726)

giansta replied:
Hi @jori-nordic, thanks a lot for finding my bug!
I tired the fix and it works. The changes I'm doing now are:
1. Remove the added callback in host and use only `BT_HCI_VS_EVT_USER`.
2. In the new callback vs_scanned, I'd remove the lines starting from calling `bt_le_adv_lookup_legacy`, as they don't seem to add anything, having the same info printed in the 2nd `printk`. This would even not require to make public the four used functions. But I'd like to have your opinion.
3. I consider that the read value rssi -127 doesn't depend on this PR changes, do you agree?

jori-nordic replied:
Good to hear!
1. Perfect, I'm fine with that.
2. Don't think you need them unless you care about (host-based) private address resolution. I propose you make the PR without those lines and open another one later if you find out you needed that functionality.
3. Yeah I think it was just because I removed the power setting, you can disregard that value in the screenshot.

giansta replied:
Now all the above mentioned changed should be pushed. There are no more changes in the host, it's everything in the sample app.

jori-nordic commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,312 @@
+/* main.c - Application main entry point */
```
Can you add a babblesim test for this sample?
@cvinayak shows how to do that in the [adv/chain test](https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/bsim/bluetooth/host/adv/chain/CMakeLists.txt#L8)

giansta replied:
I have no experience with babblesim, I tried adding the `bsim.sh` and the change in `observer.c` like in your commit, then I got an error like this:
`* scan/samples/bluetooth/hci_vs_scan_req/bsim.sh missing copyright`
As I couldn't find another case using `bsim.sh` in the tree, I gave up.

giansta commented:
> > Handle vendor specific Scan Request Received Event and provide a callback to get application events.
> > A sample application shows usage as an alternative to extended advertisements saving precious 5.5 kB RAM.
> > This PR depends on previous #66529
> 
> @giansta With the sample in this PR, with `CONFIG_BT_PERIPHERAL=y` and Dynamic Pwr Control disabled: ![image](https://private-user-images.githubusercontent.com/6350656/310084583-6a3df8cf-9fa2-4194-8eaf-743c322fb21f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk2NjE4OTgsIm5iZiI6MTcwOTY2MTU5OCwicGF0aCI6Ii82MzUwNjU2LzMxMDA4NDU4My02YTNkZjhjZi05ZmEyLTQxOTQtOGVhZi03NDNjMzIyZmIyMWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDVUMTc1OTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjFhMjZlMzFhYjY4MGZlMDY5NGIyNTU5MzhjMmRiNjFhMTU4MmExMzVhYTEwNjBlMTFhZWFkNDFlZjU3YTc5MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.LvXWCq7l0nU5mWhTXazgwNQyVjME5y5Mnc3Z5BFESqQ)
> 
> I do not see your upto ~5.5 KB saving in RAM... is your saving in your product application?

You're right, I didn't notice that the gain in RAM usage is much less in the sample app than in our product application, where we added the broadcaster role in an existing BLE central. The broadcaster role is used for debug or field operations.
Building the sample application with broadcaster role using VS Scan Request in place of extended advertisements, I get now +1508 bytes RAM; doing the same with peripheral role, the gain is +1636 bytes.
With our application, the current gains I have with same conditions are respectively +5204 and +5396 bytes RAM. When building with previous Zephyr 3.2, the gain was a bit higher (+5672).
In general, building our product application, with Zephyr 3.6, I see a slightly reduced usage both of Flash (−25776) and RAM (−692).
I don't know how these configurations changes for the broadcaster role affect the usage of RAM in other parts like the central role. If I find some time, I can try analyzing which modules got major reductions. At moment I'm happy with that. I'm updating the PR description to be more precise.

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+	BT_DATA_BYTES(BT_DATA_UUID128_ALL, 0),
+};
+
+#if defined(CONFIG_BT_EXT_ADV)
+static struct bt_le_ext_adv *ext_adv;
+#endif
```
If the purpose of this sample is to showcase the VS specific functionality, I would just omit everything related to extended advertising to make it (way) simpler

giansta replied:
The different configurations have been left to experiment with them and compare performance and memory usage, mainly building with extended advertisements compared with VS scan and with broadcaster versus peripheral role.

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+
+static void scanned(struct bt_le_ext_adv_scanned_info *info, int8_t rssi)
+{
+	char address[BT_ADDR_LE_STR_LEN] = "fake_addr";
```
```suggestion
	char address[BT_ADDR_LE_STR_LEN];
```

giansta replied:
Removed

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+}
+
+static const struct bt_le_ext_adv_cb ext_adv_callbacks = {
+	.sent = NULL,
+	.connected = NULL,
+	.scanned = scanned,
```
```suggestion
	.scanned = scanned,
```

giansta replied:
Done

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+	printk("%s (address %s rssi %d)\n", __func__, address, rssi);
+}
+
+static const struct bt_le_vs_cb  vs_callbacks = {
```
```suggestion
static const struct bt_le_vs_cb vs_callbacks = {
```

giansta replied:
Removed.

Thalley commented on include/zephyr/bluetooth/addr.h:
```diff
@@ -288,6 +288,12 @@ int bt_addr_from_str(const char *str, bt_addr_t *addr);
  */
 int bt_addr_le_from_str(const char *str, const char *type, bt_addr_le_t *addr);
 
+void bt_addr_le_copy_resolved(bt_addr_le_t *dst, const bt_addr_le_t *src);
+
+bool bt_addr_le_is_resolved(const bt_addr_le_t *addr);
+
+const bt_addr_le_t *bt_lookup_id_addr(uint8_t id, const bt_addr_le_t *addr);
+
```
These should be properly documented, and this change should be a separate commit.

Thalley replied:
Why were these moved to the public header btw?

jori-nordic replied:
I think this is just a wip push:
https://github.com/zephyrproject-rtos/zephyr/pull/68659#discussion_r1517510533

giansta replied:
Yes, removed.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2438,6 +2438,12 @@ static const struct event_handler vs_events[] = {
 	EVENT_HANDLER(BT_HCI_EVT_VS_LE_CONNECTION_IQ_REPORT, bt_hci_le_vs_df_connection_iq_report,
 		      sizeof(struct bt_hci_evt_vs_le_connection_iq_report)),
 #endif /* CONFIG_BT_DF_VS_CONN_IQ_REPORT_16_BITS_IQ_SAMPLES */
+#if defined(CONFIG_BT_CTLR_VS_SCAN_REQ_RX)
+#if defined(CONFIG_BT_BROADCASTER)
+	EVENT_HANDLER(BT_HCI_EVT_VS_SCAN_REQ_RX, bt_hci_le_vs_scan_req_received,
+		      sizeof(struct bt_hci_evt_vs_scan_req_rx)),
+#endif /* CONFIG_BT_BROADCASTER */
+#endif /* CONFIG_BT_CTLR_VS_SCAN_REQ_RX */
```
Shouldn't `CONFIG_BT_CTLR_VS_SCAN_REQ_RX` depend on `CONFIG_BT_BROADCASTER`?

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -2283,3 +2284,47 @@ void bt_hci_le_scan_req_received(struct net_buf *buf)
+
+static const struct bt_le_vs_cb *vs_cb;
+
+int bt_le_vs_scan_req_cb(const struct bt_le_vs_cb *cb)
+{
+	vs_cb = cb;
+	return 0;
+}
```
Did you consider making the callbacks a slist similar to the existing scan callbacks?

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -2283,3 +2284,47 @@ void bt_hci_le_scan_req_received(struct net_buf *buf)
+	evt = (void *)buf->data;
+	adv = adv_get_legacy();
+
+	LOG_DBG("%s peer %s rssi %d", __func__, bt_addr_le_str(&evt->addr), evt->rssi);
```
```suggestion
	LOG_DBG("peer %s rssi %d", bt_addr_le_str(&evt->addr), evt->rssi);
```

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -2283,3 +2284,47 @@ void bt_hci_le_scan_req_received(struct net_buf *buf)
+void bt_hci_le_vs_scan_req_received(struct net_buf *buf)
+{
+	struct bt_hci_evt_vs_scan_req_rx *evt;
+	struct bt_le_ext_adv *adv;
+
+	evt = (void *)buf->data;
+	adv = adv_get_legacy();
+
+	LOG_DBG("%s peer %s rssi %d", __func__, bt_addr_le_str(&evt->addr), evt->rssi);
+
+	if (!adv) {
+		LOG_ERR("No valid adv");
+		return;
+	}
```
```suggestion
	evt = (void *)buf->data;

	LOG_DBG("%s peer %s rssi %d", __func__, bt_addr_le_str(&evt->addr), evt->rssi);

	if (adv_get_legacy() == NULL) {
		LOG_ERR("No valid adv");
		return;
	}
```

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -2283,3 +2284,47 @@ void bt_hci_le_scan_req_received(struct net_buf *buf)
+		} else {
+			bt_addr_le_copy(&id_addr,
+					bt_lookup_id_addr(adv->id, &evt->addr));
+		}
+		info.addr = &id_addr;
```
```suggestion
		}
		
		info.addr = &id_addr;
```

Thalley reviewed (changes requested):
Generally I would not be in favor of this feature. 

The memory savings are not as great as claimed where my tests show less than 1 kB for enabling EXT_ADV with minimum settings, and with the additional memory usage of this, then the savings are probably less than 900B of RAM. 

I think the (small) advantage of this implementation is not worth polluting the host with this vendor specific functionality. 

I will not block it from being merged, but it won't get my approval unless a better argument than a small RAM saving is provided. 

If it indeed should get merged, then the comments from this review should be taken into account.

giansta committed:
Samples: Bluetooth: Vendor Specific: support for HCI Scan Request

Example of usage of HCI vendor specific Set Scan Request Reports
command, Scan Request Received Event and corresponding callback
to get application events. The sample application shows usage as an
alternative to extended advertisent saving precious RAM.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

giansta commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -2637,6 +2637,35 @@ int bt_le_per_adv_set_response_data(struct bt_le_per_adv_sync *per_adv_sync,
+	 * @param info Information about the scanned event.
+	 * @param rssi Received Signal Strength Indicator
+	 */
+	void (*scanned)(struct bt_le_ext_adv_scanned_info *info, int8_t rssi);
```
No more required, this change was removed.

giansta commented on samples/bluetooth/hci_vs_scan_req/README.rst:
```diff
@@ -0,0 +1,30 @@
+scan equest events even using legacy advertisements, yet saving ~ 5.5 kB RAM,
+which is quite important when the broadcaster role is added to the central role
+and RAM usage is critical.
+This sample implement only the broadcaster role, than can use either extended
```
Fixed, thanks.

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+#endif
+
+/* Lower the value, closer the scanner. */
+#define ADV_TX_POWER RADIO_TXPOWER_TXPOWER_Neg8dBm
```
It was taken from HAL header files. Are there better alternatives?

giansta commented on samples/bluetooth/hci_vs_scan_req/README.rst:
```diff
@@ -0,0 +1,30 @@
+.. _bluetooth-hci-pwr-ctrl-sample:
```
Done, thanks.

giansta commented on samples/bluetooth/hci_vs_scan_req/README.rst:
```diff
@@ -0,0 +1,30 @@
+********
+
+This simple application is a usage example to manage HCI VS commands to obtain
+scan equest events even using legacy advertisements, yet saving ~ 5.5 kB RAM,
```
I forgot to update this file, I had done only in the source code. Fixed, thanks.

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+#include <zephyr/bluetooth/hci_vs.h>
+#include <zephyr/bluetooth/addr.h>
+#include <host/hci_core.h>
+#include <common/bt_str.h>
```
Removed, thanks. Also, usage of functions defined there is now removed. of course.

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+ * (as stated in Bluetooth Core Spec 5.2, section 7.8.5)
+ */
+#define ADV_MIN_INTERVAL_625MS  BT_GAP_ADV_SLOW_INT_MIN
+#define ADV_MAX_INTERVAL_625MS  BT_GAP_ADV_SLOW_INT_MAX
```
The meaning of this was to have units in the symbols. Normally I use something like `ADV_MIN_INTERVAL_MS,` if the units are milliseconds. I agree that in this case may be unclear, removed.

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+
+static const struct bt_data scan_response_data[] = {
+	BT_DATA_BYTES(BT_DATA_UUID128_ALL, 0),
+};
```
Fixed, thanks.

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+/* BLE standard doesn't allow the scan request event with legacy advertisements.
+ * Ref: Bluetooth Core Specification v5.3, section 7.7.65.19 "LE Scan Request Received event" :
+ *      "This event shall only be generated if advertising was enabled using the
+ *       HCI_LE_Set_Extended_Advertising_Enable command."
```
Done.

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+}
+
+#if defined(CONFIG_BT_CTLR_VS_SCAN_REQ_RX)
+/* BLE standard doesn't allow the scan request event with legacy advertisements.
```
Done

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,353 @@
+ *       HCI_LE_Set_Extended_Advertising_Enable command."
+ * Added a Vendor Specific command to add this feature and save ~ 1.5 kB SRAM.
+ * (Higher RAM saving up to 5 kB observed in larger applications).
+ */
```
Done

giansta commented:
> Generally I would not be in favor of this feature.
> 
> The memory savings are not as great as claimed where my tests show less than 1 kB for enabling EXT_ADV with minimum settings, and with the additional memory usage of this, then the savings are probably less than 900B of RAM.
> 
> I think the (small) advantage of this implementation is not worth polluting the host with this vendor specific functionality.
> 
> I will not block it from being merged, but it won't get my approval unless a better argument than a small RAM saving is provided.
> 
> If it indeed should get merged, then the comments from this review should be taken into account.

Thanks, @Thalley for all your inputs.

About memory saving, in my [previous reply](https://github.com/zephyrproject-rtos/zephyr/pull/68659#issuecomment-1980487358) to @jori-nordic, I gave some details about the memory savings when building for my nRF52832 target: slightly more 1.5 kB for the sample app can be considered not much, I agree; but in our application we currently save still more than 5 kB (they were more than 5.5 kB when using Zephyr 3.2, v3.6 had quite good optimizations).

In our nRF52832 based board saving 5 of 64 kB isn't negligible. We were using already great part of them, we did some good optimizations, and this is my contribution in having a new feature with almost no additional RAM to help in field activities during installation and for triggering connection to our backend. Using VS commands looks really suitable for our use case, as @jori-nordic suggested since beginning.

Moreover, thanks to his latest suggestions, the changes in the host disappeared and everything is now only in the sample app (and in our application).

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,326 @@
+#include <zephyr/bluetooth/addr.h>
+
+#if !defined(CONFIG_BT_EXT_ADV) && !defined(CONFIG_BT_CTLR_VS_SCAN_REQ_RX)
+#error "Either BT_EXT_ADV or BT_CTLR_VS_SCAN_REQ_RX must be selected to have scan request callback"
```
The run test with `qemu_cortex_m3` is triggering this `#error `directive. This seems due to the fact that some configs, including `CONFIG_BT_CTLR_VS_SCAN_REQ_RX`, have been moved in `overlay-bt_ll_sw_split.conf`.
@jori-nordic, do you have any idea about how to make qemu aware of it? Or whatever other fix? (Better not removing this check).

giansta replied:
> In this particular case, the CI is right. The qemu target compiles with an "off-chip" controller. E.g. over UART. In that case, all the `BT_CTLR` kconfigs are not present. So you can't really assert on them.
> 
> To fix that, you can:
> 
>     * add an early-exit in `main()` if `BT_EXT_ADV` is not present
> 
>     * add error handling + log msg [here](https://github.com/zephyrproject-rtos/zephyr/blob/cf40a7f67519fdd3d2576c33261d08962a947829/samples/bluetooth/hci_vs_scan_req/src/main.c#L136) instructing the user what config to enable on the controller board.
> 
> 
> edit: On re-reading the sample, I would prefer if it was only demonstrating the VS command and not the extended advertising part. It would simplify the code much more and be more focused.

At the end, it was easier to remove these checks. I have similar ones in our application' module managing the advertisements, not really needed in sample app.

giansta commented:
> > > Generally I would not be in favor of this feature.
> > > The memory savings are not as great as claimed where my tests show less than 1 kB for enabling EXT_ADV with minimum settings, and with the additional memory usage of this, then the savings are probably less than 900B of RAM.
> > > I think the (small) advantage of this implementation is not worth polluting the host with this vendor specific functionality.
> > > I will not block it from being merged, but it won't get my approval unless a better argument than a small RAM saving is provided.
> > > If it indeed should get merged, then the comments from this review should be taken into account.
> > 
> > 
> > Thanks, @Thalley for all your inputs.
> > About memory saving, in my [previous reply](https://github.com/zephyrproject-rtos/zephyr/pull/68659#issuecomment-1980487358) to @jori-nordic, I gave some details about the memory savings when building for my nRF52832 target: slightly more 1.5 kB for the sample app can be considered not much, I agree; but in our application we currently save still more than 5 kB (they were more than 5.5 kB when using Zephyr 3.2, v3.6 had quite good optimizations).
> > In our nRF52832 based board saving 5 of 64 kB isn't negligible. We were using already great part of them, we did some good optimizations, and this is my contribution in having a new feature with almost no additional RAM to help in field activities during installation and to wake up these devices. Using VS commands looks really suitable for our use case, as @jori-nordic suggested since beginning.
> > Moreover, thanks to his latest suggestions, the changes in the host disappeared and everything is now only in the sample app (and in our application).
> 
> Thanks for your reply :)
> 
> Specific savings will always depend on your configuration, and for the nRF52840 that I did my tests on, the savings would be less than 1 kB using the newest main for the samples, but of course this depends on the Kconfig values as well as the application (Which I why I wanted to get any specific numbers be removed).
> 
> With everything in the sample now (thank you!), I'm no longer against this :) Given that this is a sample, and that samples should, IMO, be as simple as possible (and not defined to be used for performance experimentation), I would still suggest to remove the EXT_ADV stuff from it, so that it be much clearer what it does, and be a lot easier to read. If you do want to keep the EXT_ADV stuff, then I would suggest to modify the code so that everything related to one is in a single block (or even in a single file), and ditto for the other, although I'm still in favor of just removing everything related to EXT_ADV as this VS command is specifically for devices that does not want to use that, right?

I've both removed specific values of memory saving and all code using extended adv.

giansta commented on samples/bluetooth/hci_vs_scan_req/README.rst:
```diff
@@ -0,0 +1,30 @@
+This simple application is a usage example to manage HCI VS commands to obtain
+scan equest events even using legacy advertisements, yet saving ~ 1.5 kB RAM.
+This is quite important in applications in which the broadcaster role is added
+to the central role, where the RAM saving is up to 5 kB.
```
Done

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,326 @@
+
+static const struct bt_data scan_response_data[] = {
+	BT_DATA(BT_DATA_NAME_COMPLETE, scan_data, sizeof(scan_data)),
+};
```
Not sure to understand this request. In other samples I see same scan response data with device name only. Do you mean adding flags, like done now?

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,326 @@
+	.interval_max = ADV_MAX_INTERVAL,
+};
+
+static const struct bt_data advertisement_data[] = {
```
Done

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,326 @@
+#if defined(CONFIG_BT_CTLR_VS_SCAN_REQ_RX)
+	err = bt_hci_register_vnd_evt_cb(vs_scanned);
+	enable_legacy_adv_scan_request_event(true);
+	err = bt_le_adv_start(&parameters, advertisement_data, ARRAY_SIZE(advertisement_data),
```
Done, thanks.

giansta commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,326 @@
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
```
Indeed in our application the equivalent module does exactly as you say. I lost the guard when porting part of that code in the sample app, even because the order is different due to different coding rules. Fixed now.

giansta commented:
The BabbleSim Tests  seem to have errors somewhere else, right?

jori-nordic commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,228 @@
+/* Allow different TX power per role/connection. It requires CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL.
+ * same as in samples/bluetooth/hci_pwr_ctrl/src
+ */
+static void set_tx_power(uint8_t handle_type, uint16_t handle, int8_t tx_pwr_lvl)
```
Does this sample really need to set the TX power?
I think you should remove it if not absolutely necessary, as to make the sample as focused and streamlined as possible.

giansta replied:
It's necessary in our app module, not in this sample: removed.

jori-nordic reviewed (changes requested):
No further comments. The rest of the PR is ok for me

Thalley commented on samples/bluetooth/hci_vs_scan_req/README.rst:
```diff
@@ -0,0 +1,29 @@
+********
+
+This simple application is a usage example to manage HCI VS commands to obtain
+scan equest events even using legacy advertisements, while saving RAM.
+This is quite important in applications in which the broadcaster role is added
+to the central role, where the RAM saving can be bigger.
```
```suggestion
scan equest events even using legacy advertisements, while may result in lower RAM usage than using extended advertising.
This is quite important in applications in which the broadcaster role is added
to the central role, where the RAM saving can be bigger.
```

giansta replied:
Done

Thalley commented on samples/bluetooth/hci_vs_scan_req/overlay-bt_ll_sw_split.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_BT_CTLR_VS_SCAN_REQ_RX=y
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
```
Why do we need advanced features for this sample?

giansta replied:
It's a leftover from our application module, not needed here: removed.

Thalley commented on samples/bluetooth/hci_vs_scan_req/overlay-bt_ll_sw_split.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_BT_CTLR_VS_SCAN_REQ_RX=y
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL=y
```
Why do we need dynamic power control for this sample?

giansta replied:
It's required in our app module control of Tx power per role, not in this sample: removed.

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,228 @@
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LENGTH),
+};
+
+static const struct bt_data scan_rsp_data[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA(BT_DATA_NAME_COMPLETE, scan_data, sizeof(scan_data)),
+};
```
IIRC a device may only have single name, so if you use `BT_DATA_NAME_COMPLETE` in both the ADV data and the scan response data, they shall be the same value. 

You can use e.g. `BT_DATA_MANUFACTURER_DATA` if you want to use some advertising data without any requirements

giansta replied:
Sorry for my lack of fantasy. Hope it's better now.

Thalley replied:
You may have missed something here :) 

My point is that you shouldn't have different values for `BT_DATA_NAME_COMPLETE` in the advertising and scan response data. 

So I suggest to do either

```suggestion
static const struct bt_data scan_rsp_data[] = {
	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LENGTH),
};
```

or

```suggestion
static const struct bt_data scan_rsp_data[] = {
	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
	BT_DATA(BT_DATA_MANUFACTURER_DATA, scan_data, sizeof(scan_data)),
};
```

giansta replied:
ok for the 2nd

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,228 @@
+	if (err) {
+		printk("VS user callback register err %d\n", err);
+		return err;
+	}
+	enable_legacy_adv_scan_request_event(true);
```
```suggestion
	}
	
	enable_legacy_adv_scan_request_event(true);
```

Suggest to keep an empty line after ending `}` (not a requirement)

giansta replied:
Done

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,228 @@
+{
+	printk("Disconnected (reason 0x%02x)\n", reason);
+
+	(void) start_advertising();
```
```suggestion
	(void)start_advertising();
```

giansta replied:
Done

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,228 @@
+	if (err) {
+		printk("Start legacy adv err %d\n", err);
+		return err;
+	}
+	printk("Advertising successfully started (%s)\n", CONFIG_BT_DEVICE_NAME);
+	return 0;
```
```suggestion
	}
	
	printk("Advertising successfully started (%s)\n", CONFIG_BT_DEVICE_NAME);
	
	return 0;
```

Suggest to keep empty line before `return` statements. (Not a requirement)

giansta replied:
Done

Thalley reviewed (changes requested):
A few more minor requests

jori-nordic commented on samples/bluetooth/hci_vs_scan_req/README.rst:
```diff
@@ -0,0 +1,30 @@
+RAM usage than using extended advertising.
+This is quite important in applications in which the broadcaster role is added
+to the central role, where the RAM saving can be bigger.
+This sample implements only the broadcaster role; the peripheral role with
```
A last comment and then I'll approve:
The sample can be slimmed down even more, by removing the `CONFIG_BT_PERIPHERAL` support.

I don't see it adding anything related to this VS event in the code. Users can easily copy code from peripheral samples if they need it.

Thalley replied:
I agree with this. Since the purpose of this sample is to show the VS command, then it should do just that.

giansta replied:
removed

jori-nordic commented on samples/bluetooth/hci_vs_scan_req/prj.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_LOG=y
+CONFIG_BT_BROADCASTER=y
+CONFIG_BT_DEVICE_NAME="VS Scan Notify"
+#CONFIG_BT_PERIPHERAL=y
```
remove

giansta replied:
removed

Thalley commented on samples/bluetooth/hci_vs_scan_req/overlay-bt_ll_sw_split.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_BT_CTLR_VS_SCAN_REQ_RX=y
```
Should this be moved to the prj.conf file? 

Since the main purpose of the sample is to use this, it shouldn't be in an overlay IMO

giansta replied:
It was asked above to move some configs into this overlay file. Config moved and overlay removed, if it doesn't break anything in CI.

cvinayak replied:
I was only concern if the sample was to be used with other Controllers that may decide to add the feature in which case supplying overlay files seemed fair.

I have no objections as-is to have in prj.conf, and changes can  be done in the future if required.

Thalley commented on samples/bluetooth/hci_vs_scan_req/src/main.c:
```diff
@@ -0,0 +1,184 @@
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	if (err) {
+		printk("Connection failed (err 0x%02x)\n", err);
+	} else {
+		bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+		printk("Connected at %s\n", addr);
+	}
```
Minor: 

```suggestion

	if (err) {
		printk("Connection failed (err 0x%02x)\n", err);
	} else {
		char addr[BT_ADDR_LE_STR_LEN];
		
		bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
		printk("Connected at %s\n", addr);
	}
```

giansta replied:
Code removed (CONFIG_BT_BROADCASTER).

Thalley reviewed (changes requested):
A few final thoughts, but otherwise LGTM. 

Thanks for your quick updates and patience :)

jori-nordic commented:
@cvinayak could you re-review?

giansta commented:
No way to merge this PR?

Thalley commented:
> No way to merge this PR?

@giansta You need to rebase the PR as there is a merge conflict

giansta commented:
> > No way to merge this PR?
> 
> @giansta You need to rebase the PR as there is a merge conflict

Sorry, I was focusing on the above request for review and missed the below conflict. Rebased now.

Thalley commented:
@cvinayak please dismiss your review or re-review this

giansta commented:
Can I ask if something is preventing this PR to be merged? Any further action required?
